### ISSUE TITLE
Padding side

### ISSUE THREAD

**Issue by waynechu1021** [siyan-zhao/OPSD#30]
Thank you for open-sourcing such a great piece of work. My question is, shouldn't left padding be used during generation? I see in your codebase that right padding is forced, which might cause pad tokens between the prompt and generation for some samples. Please point out if I'm wrong. Looking forward to your reply.

**Comment by yu-ki3406** [comment]
We evaluated OPSD with left padding, and the results in Appendix C show that it remains effective under this setting. The results are available here: https://arxiv.org/pdf/2608.09228

**Comment by waynechu1021** [comment]
> We evaluated OPSD with left padding, and the results in Appendix C show that it remains effective under this setting. The results are available here: https://arxiv.org/pdf/2608.09228

thanks for your reply！ I am still confused why using right padding in the roll out of the student is effective. It seems that left padding in the roll out would bring lower performance although it is the correct padding way.

**Comment by yu-ki3406** [comment]
Thanks for the follow-up! In our paper, we found that OPSD does not necessarily rely on matching the teacher’s exact answer or token-level trajectory. Instead, the student appears to acquire mathematical knowledge from the teacher, even when the teacher and student are not solving exactly the same problem.

This may help explain why right padding can still work well: even if it introduces some token-level misalignment, OPSD does not seem to require exact token correspondence between the teacher and student. In that sense, the effectiveness of right padding is consistent with our findings.

**Comment by waynechu1021** [comment]
> Thanks for the follow-up! In our paper, we found that OPSD does not necessarily rely on matching the teacher’s exact answer or token-level trajectory. Instead, the student appears to acquire mathematical knowledge from the teacher, even when the teacher and student are not solving exactly the same problem.
> 
> This may help explain why right padding can still work well: even if it introduces some token-level misalignment, OPSD does not seem to require exact token correspondence between the teacher and student. In that sense, the effectiveness of right padding is consistent with our findings.

I see! That's a great work!

**Comment by siyan-zhao** [comment]
hi, thanks for catching this. and @yu-ki3406 thanks for sharing your paper and the reproduced results of OPSD.

i think you're right that right padding affects some examples, but here it only affects the first generated token: the logits used to sample it are computed at the padding token's position (since generate() reads the last position in the sequence for the next-token logits), so that first token can come out wrong. after that, since the attention mask excludes the padding tokens from being attended to (and position ids are computed to correctly skip the padding gap for later tokens), generation is done correctly now. so the impact of this issue is likely limited, which may help explain why OPSD still improves scores and as with what @yu-ki3406 explained above as well. Left padding would be more accurate.
